lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+        &self.virtio_state.pci_state
+    }
+}
+
+impl VirtioDevice for PciVirtioSoftNpuPort {
+    fn cfg_rw(&self, mut rwo: RWOp) {
+        NET_DEV_REGS.process(&mut rwo, |id, rwo| match rwo {
+            RWOp::Read(ro) => self.net_cfg_read(id, ro),
+            RWOp::Write(_) => {
+                //ignore writes
+            }
+        });
+    }
+
+    fn get_features(&self) -> u32 {
+        VIRTIO_NET_F_MAC
+    }
+
+    fn set_features(&self, _feat: u32) {}
+
+    fn queue_notify(&self, vq: &Arc<VirtQueue>) {
+        self.handle_guest_virtio_request(vq);
+    }
+}
+
+struct PacketHandler {}
+
+impl PacketHandler {
+    /// Spawn a thread that handles packets coming into the emulated ASIC for
+    /// the specified interface index.
+    fn run_ingress_packet_handler_thread(
+        index: usize,
+        data_handles: Vec<dlpi::DlpiHandle>,
+        virtio: Arc<PortVirtioState>,
+        pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
+        log: Logger,
+    ) {
+        spawn(move || {
+            info!(log, "ingress packet handler is running for port {}", index,);
+            Self::run_ingress_packet_handler(
+                index,
+                data_handles,
+                virtio.clone(),
+                pipeline.clone(),
+                log,
+            )
+        });
+    }
+
+    /// Handle packets coming into the emulated ASIC for the specified interface
+    /// index.
+    fn run_ingress_packet_handler(
+        index: usize,
+        data_handles: Vec<dlpi::DlpiHandle>,
+        virtio: Arc<PortVirtioState>,
+        pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
+        log: Logger,
+    ) {
+        let dh = data_handles[index];
+        loop {
+            //
+            // wait for a packet from dlpi
+            //
+            let mut src = [0u8; dlpi::sys::DLPI_PHYSADDR_MAX];
+            let mut msg = [0u8; MTU];
+            let mut recvinfo = dlpi_recvinfo_t::default();
+            let n = match dlpi::recv(
+                dh,
+                &mut src,
+                &mut msg,
+                -1, // block until we get something
+                Some(&mut recvinfo),
+            ) {
+                Ok((_, n)) => n,
+                Err(e) => {
+                    error!(log, "rx error at index {}: {}", index, e);
+                    continue;
+                }
+            };
+
+            //
+            // process packet with loaded P4 program
+            //
+
+            // TODO pipeline should not need to be mutable for packet handling?
+            let pkt = packet_in::new(&msg[..n]);
+            let mut p = pipeline.lock().unwrap();
+            let pl = match &mut *p {
+                Some(ref mut pl) => &mut pl.1,
+                None => continue, // no program is loaded
+            };
+
+            Self::process_external_packet(
+                index + 1,
+                pkt,
+                &data_handles,
+                virtio.clone(),
```

Comment:
Any particular reason why the virtio refs are being cloned for every packet?

Reply:
Nope